drivers/video/fbdev/bcm2708_fb.c

```diff
@@ -883,87 +1034,149 @@ static int bcm2708_fb_probe(struct platform_device *dev)
+	 */
+	if (!fbdev->cb_base) {
+		fbdev->cb_base = dma_alloc_writecombine(&dev->dev, SZ_64K,
+							&fbdev->cb_handle,
+							GFP_KERNEL);
+		if (!fbdev->cb_base) {
+			dev_err(&dev->dev, "cannot allocate DMA CBs\n");
+			ret = -ENOMEM;
+			goto free_fb;
+		}
 
-	pr_info("BCM2708FB: allocated DMA memory %pad\n", &fb->cb_handle);
+		dev_info(&dev->dev,
+			 "Allocated DMA memory %08x\n",
+			 fbdev->cb_handle);
 
-	ret = bcm_dma_chan_alloc(BCM_DMA_FEATURE_BULK,
-				 &fb->dma_chan_base, &fb->dma_irq);
-	if (ret < 0) {
-		dev_err(&dev->dev, "couldn't allocate a DMA channel\n");
-		goto free_cb;
+		ret = bcm_dma_chan_alloc(BCM_DMA_FEATURE_BULK,
+					 &fbdev->dma_chan_base,
+					 &fbdev->dma_irq);
+		if (ret < 0) {
+			dev_err(&dev->dev, "Couldn't allocate a DMA channel\n");
+			goto free_cb;
+		}
+		fbdev->dma_chan = ret;
+
+		ret = request_irq(fbdev->dma_irq, bcm2708_fb_dma_irq,
+				  0, "bcm2708_fb DMA", fbdev);
+		if (ret) {
+			dev_err(&dev->dev,
+				"Failed to request DMA irq\n");
+			goto free_dma_chan;
+		}
+		dev_info(&dev->dev,
+			 "Allocated DMA channel %d @ %p\n",
+			 fbdev->dma_chan, fbdev->dma_chan_base);
 	}
-	fb->dma_chan = ret;
 
-	ret = request_irq(fb->dma_irq, bcm2708_fb_dma_irq,
-			  0, "bcm2708_fb dma", fb);
-	if (ret) {
-		pr_err("%s: failed to request DMA irq\n", __func__);
-		goto free_dma_chan;
-	}
+	for (i = 0; i < num_displays && ret == 0; i++) {
+		struct bcm2708_fb *fb = &fbdev->displays[i];
 
-	pr_info("BCM2708FB: allocated DMA channel %d\n", fb->dma_chan);
+		fb->fw = fw;
+		fb->display_settings.display_num = i;
 
-	fb->dev = dev;
-	fb->fb.device = &dev->dev;
+		fb->dev = dev;
+		fb->fb.device = &dev->dev;
+		/* failure here isn't fatal, but we'll fail in vc_mem_copy
+		 * if fb->gpu is not valid
+		 */
+		rpi_firmware_property(fb->fw,
+				      RPI_FIRMWARE_GET_VC_MEMORY,
+				      &fb->gpu, sizeof(fb->gpu));
 
-	/* failure here isn't fatal, but we'll fail in vc_mem_copy if
-	 * fb->gpu is not valid
-	 */
-	rpi_firmware_property(fb->fw, RPI_FIRMWARE_GET_VC_MEMORY, &fb->gpu,
-			      sizeof(fb->gpu));
+		if (old) {
+			memset(&fb->display_settings, 0,
+			       sizeof(fb->display_settings));
+		} else {
+			fb->display_settings.display_num = i;
+			ret = rpi_firmware_property(fb->fw,
+						    RPI_FIRMWARE_FRAMEBUFFER_GET_DISPLAY_SETTINGS,
+						    &fb->display_settings,
+						    GET_DISPLAY_SETTINGS_PAYLOAD_SIZE);
+		}
 
-	ret = bcm2708_fb_register(fb);
-	if (ret == 0) {
-		platform_set_drvdata(dev, fb);
-		goto out;
+		ret = bcm2708_fb_register(fb);
+
+		if (ret == 0) {
+			fb->fbdev = fbdev;
+			bcm2708_fb_debugfs_init(fb);
+
+			fbdev->num_displays++;
+
+			dev_info(&dev->dev,
+				 "Registered framebuffer for display %d, size %dx%d\n",
+				 fb->display_settings.display_num,
+				 fb->fb.var.xres,
+				 fb->fb.var.yres);
+		} else {
+			goto free_dma_chan; // TODO. Better cleanup
+		}
 	}
 
+	init_waitqueue_head(&fbdev->dma_waitq);
+
+	platform_set_drvdata(dev, fbdev);
+
+	return ret;
+
```

Comment:
If fb1 fails during creation, you haven't destroyed fb0 before quiting.

Reply:
Hm, if fb0 was OK should probably let that remain working so something is displayed. Will need to sort out an error path for that use case.

Reply:
OK, will now continue to loop, but will shut down if NO framebuffers are instantiated, otherwise will simply continue with those that succeeded.